bugsnag/src/main/java/com/bugsnag/callbacks/DeviceCallback.java

```diff
@@ -39,10 +48,24 @@ private static String lookupHostname() {
             return hostname;
         }
 
-        // Resort to dns hostname lookup
+        // Resort to dns hostname lookup.
+        // Look up the hostname in a daemon thread, with a timeout of HOSTNAME_LOOKUP_TIMEOUT.
+        // This is to prevent applications hanging waiting for the dns lookup to resolve
+        ExecutorService lookupService = Executors.newSingleThreadExecutor(new DaemonThreadFactory());
+        Future<String> future = lookupService.submit(new Callable<String>() {
+            @Override
+            public String call() throws UnknownHostException {
+                return InetAddress.getLocalHost().getHostName();
+            }
+        });
```

Comment:
Instead of adding a timeout to the `lookupHostname` method, maybe we should instead wrap the "startup" cache in a daemon thread?

```java
    public static void initializeCache() {
        Thread hostnameLookup = new Thread("Hostname Lookup") {
            @Override
            public void run() {
                getHostnameValue();
            }
        };
        hostnameLookup.setDaemon(true);
        hostnameLookup.start();
    }
```

Reply:
As discussed, I've moved the cache init to a daemon thread to prevent blocking on initialization, but retained the timeout on the hostname lookup as `getHostnameValue` may also be called from elsewhere before the cache is initialized.